### PR TITLE
Add support for ROS 2 (Humble)

### PR DESCRIPTION
**Pull request by HansRobo**
## abstruct

I imported derverbles from choreonoid/choreonoid_ros2 and added some fixes.
With this pull-request, this repository is configured to be buildable on both ROS 1 and ROS 2.

The major differences between ROS 1 version and ROS 2 version are as follows.

### ROS 1

- A choreonoid app containing a ROS plugin behaves as a single ROS context, a node.
- The plugin initialization blocks the choreonoid app launch if there is no connection to roscore.

### ROS 2

- Since there is no ROS 2-related processing in plugin initialization, it does not block the app launch.
- ROS 2 plugin can be loaded and used in choreonoid app without modification for the plugin.
- Each of WorldROS2Item and BodyROS2Item operates as its own ROS 2 context and node for each instance of the item.

### ROS 1

ros_control support is available now.

### ROS 2

ros_control support is not available now.
However, we plan to support it in the future.

## Build instruction for ROS 2 (Humble, Ubuntu22.04)

```bash
mkdir -p cnoid_ws/src
cd cnoid_ws/src
git clone git@github.com:choreonoid/choreonoid.git
git clone -b merge_choreonoid_ros2 git@github.com:choreonoid/choreonoid_ros.git
cd ..
source /opt/ros/humble/setup.bash
rosdep install -iy --from-paths src
colcon build --symlink-install
```

Note: Please ignore the error from choreonoid with catkin, when running rosdep (Fixed in https://github.com/choreonoid/choreonoid/pull/36)

### CODE REVIEW

File: src/node/choreonoid_ros2.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{
+    cnoid::App app(argc, argv, "Choreonoid-ROS2", "Choreonoid");
+
+    auto plugin_manager = cnoid::PluginManager::instance();
+
+    if(auto pluginPath = getenv("CNOID_PLUGIN_PATH")){
+      plugin_manager->addPluginPath(toUTF8(pluginPath));
+    }
+
+    auto plugin_path = ament_index_cpp::get_package_share_directory("choreonoid_ros2") + "/../../lib/choreonoid_ros2";
```
Comment (ssr-yuki) [inline src/node/choreonoid_ros2.cpp]: In my environment, this line does not work: executing Choreonoid with `ros2 run` fails. Should we replace "choreonoid_ros2" into "choreonoid_ros"?

File: src/node/choreonoid_ros2.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include <cnoid/PluginManager>
+#include <cnoid/ProjectManager>
+#include <cnoid/UTF8>
+#include <cstdlib>
+#include <rclcpp/rclcpp.hpp>
+#include <regex>
+
+using namespace std;
+using namespace cnoid;
+namespace filesystem = stdx::filesystem;
```
Comment (ssr-yuki) [inline src/node/choreonoid_ros2.cpp]: Viewing the whole of this file, I guess this lines are unnecessary. Please let me know if you use this anywhere.

I would also like to point out that this file explicitly write `cnoid::` in some lines while it is `using namespace cnoid`. Could you unify the codes into either style? (I personally prefer not to use `using namespace` directives.)

File: src/plugin/BodyROS2Item.cpp [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+
+
+Item *BodyROS2Item::doDuplicate() const
+{
+    return new BodyROS2Item(*this);
+}
+
+bool BodyROS2Item::store(Archive &archive)
+{
+    archive.write("body_ros_version", 0);
```
Comment (ssr-yuki) [inline src/plugin/BodyROS2Item.cpp:60]: The parameter `body_ros_version` is no longer used as not restored in the next function.

File: src/plugin/BodyROS2Item.cpp [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+
+#include <geometry_msgs/msg/point32.hpp>
+#include <geometry_msgs/msg/wrench_stamped.hpp>
+#include <sensor_msgs/image_encodings.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <sensor_msgs/msg/imu.hpp>
+#include <sensor_msgs/msg/laser_scan.hpp>
+
+#include <sensor_msgs/msg/point_cloud2.hpp>
+typedef sensor_msgs::msg::PointCloud2 PointCloudTypeForRangeSensor;
```
Comment (ssr-yuki) [inline src/plugin/BodyROS2Item.cpp]: Please delete this typedef, which is no longer used. The empty line L14 is also removable.

File: src/plugin/BodyROS2Item.cpp [hunk 5]
```diff
@@ -0,0 +1,687 @@
...
+        jointState.position[i] = joint->q();
+        jointState.velocity[i] = joint->dq();
+        jointState.effort[i] = joint->u();
+    }
+
+    createSensors(simulationBody);
+
+    jointStatePublisher
+        = rosNode->create_publisher<sensor_msgs::msg::JointState>(getROS2Name("joint_states"),
+                                                                  1000);
```
Comment (ssr-yuki) [inline src/plugin/BodyROS2Item.cpp:133]: just my interest: is 1000 a good choice in ROS 2 publication?

File: src/plugin/WorldROS2Item.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#ifndef CNOID_ROS_PLUGIN_WORLD_ROS_ITEM_H
+#define CNOID_ROS_PLUGIN_WORLD_ROS_ITEM_H
```
Comment (ssr-yuki) [inline src/plugin/WorldROS2Item.h]: should be `~_ROS2_ITEM_H`.

**Review by ssr-yuki — CHANGES_REQUESTED** [review]
Thank you for your fantastic development. I leave some comments for improvement and clarification of the codes.

While we still need to work for further implementation, we now get a foot in the door of ROS 2 with your PR!

File: src/node/choreonoid_ros2.cpp (hunk 1 above)
Comment (HansRobo) [reply src/node/choreonoid_ros2.cpp]: Fixed in https://github.com/choreonoid/choreonoid_ros/pull/18/commits/2af846c245b1e791a3d3f73631c15a8d2c57b242

File: src/plugin/WorldROS2Item.h (hunk 6 above)
Comment (HansRobo) [reply src/plugin/WorldROS2Item.h]: Fixed in https://github.com/choreonoid/choreonoid_ros/pull/18/commits/54d8bc3b13d89ebebcb4b816785a2a907d14c989

File: src/plugin/BodyROS2Item.cpp (hunk 4 above)
Comment (HansRobo) [reply src/plugin/BodyROS2Item.cpp]: Fixed in https://github.com/choreonoid/choreonoid_ros/pull/18/commits/54d8bc3b13d89ebebcb4b816785a2a907d14c989

File: src/node/choreonoid_ros2.cpp (hunk 2 above)
Comment (HansRobo) [reply src/node/choreonoid_ros2.cpp]: Fixed in https://github.com/choreonoid/choreonoid_ros/pull/18/commits/1f681d773093b319e364a62a441848a438f70f4f

File: src/plugin/BodyROS2Item.cpp (hunk 5 above)
Comment (HansRobo) [reply src/plugin/BodyROS2Item.cpp:133]: The history depth for publisher should be comprehensively judged based on the cycle of publication, the importance of the content, the environment, etc.
I don't think 1000 (which can withstand 10 seconds of communication loss) is appropriate.
The old joint state has almost no value, so if a communication breakdown occurs, the old data should be dropped and new data published.
This is the same between ROS 1 and 2.

**Review by ssr-yuki — APPROVED** [review]
Thank you very much!

**Pull request merged**